pkg/auth/dependency/sso/apple.go

```diff
@@ -0,0 +1,207 @@
+		oauthConfig:    f.OAuthConfig,
+		urlPrefix:      f.URLPrefix,
+		providerConfig: f.ProviderConfig,
+		encodedState:   encodedState,
+		baseURL:        "https://appleid.apple.com/auth/authorize",
+		responseMode:   "form_post",
+		nonce:          state.Nonce,
+	}
+	return authURL(p)
+}
+
+func (f *AppleImpl) GetAuthInfo(r OAuthAuthorizationResponse, state State) (authInfo AuthInfo, err error) {
+	return f.OpenIDConnectGetAuthInfo(r, state)
+}
+
+func (f *AppleImpl) OpenIDConnectGetAuthInfo(r OAuthAuthorizationResponse, state State) (authInfo AuthInfo, err error) {
+	if subtle.ConstantTimeCompare([]byte(state.Nonce), []byte(crypto.SHA256String(r.Nonce))) != 1 {
+		err = NewSSOFailed(InvalidParams, "invalid sso state")
+		return
+	}
+
+	keySet, err := f.getKeys()
+	if err != nil {
+		err = NewSSOFailed(NetworkFailed, "failed to get OIDC JWTs")
+		return
+	}
+
+	clientSecret, err := f.createClientSecret()
+	if err != nil {
+		err = errors.Newf("failed to create client secret: %w", err)
+		return
+	}
+
+	body := url.Values{}
+	body.Set("grant_type", "authorization_code")
+	body.Set("client_id", f.ProviderConfig.ClientID)
+	body.Set("code", r.Code)
+	body.Set("redirect_uri", redirectURI(f.URLPrefix, f.ProviderConfig))
+	body.Set("client_secret", clientSecret)
+
+	resp, err := http.PostForm("https://appleid.apple.com/auth/token", body)
+	if err != nil {
+		err = NewSSOFailed(NetworkFailed, "failed to connect authorization server")
+		return
+	}
+	defer resp.Body.Close()
+
+	var tokenResp AccessTokenResp
+	if resp.StatusCode == 200 {
+		err = json.NewDecoder(resp.Body).Decode(&tokenResp)
+		if err != nil {
+			return
+		}
+	} else {
+		var errorResp oauthErrorResp
+		err = json.NewDecoder(resp.Body).Decode(&errorResp)
+		if err != nil {
+			return
+		}
+		err = errorResp.AsError()
+		return
+	}
+
+	// https://developer.apple.com/documentation/signinwithapplerestapi/verifying_a_user
+	// The following code verify the id token according to the docs.
+
+	idToken := tokenResp.IDToken()
+	keyFunc := func(token *jwt.Token) (interface{}, error) {
+		keyID, ok := token.Header["kid"].(string)
+		if !ok {
+			return nil, NewSSOFailed(SSOUnauthorized, "no kid")
+		}
+		if key := keySet.LookupKeyID(keyID); len(key) == 1 {
+			return key[0].Materialize()
+		}
+		return nil, NewSSOFailed(SSOUnauthorized, "failed to find signing key")
+	}
+
+	mapClaims := jwt.MapClaims{}
+	// Verify the signature
+	_, err = jwt.ParseWithClaims(idToken, mapClaims, keyFunc)
+	if err != nil {
+		err = NewSSOFailed(SSOUnauthorized, "invalid JWT signature")
+		return
+	}
+
+	// Verify the nonce
+	hashedNonce, ok := mapClaims["nonce"].(string)
+	if !ok {
+		err = NewSSOFailed(InvalidParams, "no nonce")
+		return
+	}
+	if subtle.ConstantTimeCompare([]byte(hashedNonce), []byte(crypto.SHA256String(r.Nonce))) != 1 {
+		err = NewSSOFailed(SSOUnauthorized, "invalid nonce")
+		return
+	}
+
+	// Verify the issuer
+	if !mapClaims.VerifyIssuer("https://appleid.apple.com", true) {
```

Comment:
The referenced document says the issuer should 'contains' the constant string. I'm not sure if it'll be some other strings like `https://appleid.apple.com/abc` in future.

Reply:
The wording is a bit confusing. In my real testing, the value of issuer is exactly `https://appleid.apple.com`. If I change to use `strings.HasPrefix`, do you have any concerns?

Reply:
My concern is that the document explicitly used the word 'contains' for `iss` field, and the word 'is' is used for exact matching the `aud` field. It seems to me that they anticipated the need to change it in future. I think using `strings.Contains` would be safer.